### ISSUE TITLE
Virtual File System issues

### ISSUE THREAD

**Issue by thetestgame** [panda3d/panda3d#225]
Panda seems to have issues when you have complex model-path prc settings. I have a model folder structure that looks similar to this
```
- resources
  - pack1
  - pack2
  - pack3
  - pack4
  - pack5
```
 
prc:
```
model-path resources
model-path resources/pack1
model-path resources/pack2
model-path resources/pack3
model-path resources/pack4
model-path resources/pack5
```

It is split up to help make packing for distribution easier on file size. However various files will not be found by the vfs. There is also presence of an assertion relating to the directories in ``dtool/src/dSearchPath.cxx`` on line ``262``
```
n < _directories.size()
```

**Comment by rdb** [comment]
I made this test case based on your description of the issue:
```python
import os

for dir in "pack1", "pack2", "pack3", "pack4", "pack5":
    dirname = os.path.join("resources", dir)
    if not os.path.isdir(dirname):
        os.makedirs(dirname)

with open("resources/pack1/test.egg", "w") as fp:
    fp.write("<Group> test {}")

from panda3d.core import *

loadPrcFileData("", """
model-path resources
model-path resources/pack1
model-path resources/pack2
model-path resources/pack3
model-path resources/pack4
model-path resources/pack5
""")

from direct.directbase import DirectStart
node = loader.loadModel("test.egg")
assert node.find("test")
```

However, it does not exhibit the issue.  Could you give further description of the situations in which the problem occurs, or better yet, submit a test case that demonstrates the issue?

**Comment by thetestgame** [comment]
This issue only seems to happen in cases of tons of resources. I've encountered it twice. Once with my own project and also while working on a Pirates of the Caribbean Online remake. In both cases the model path had over 13,000 assets in it and several sub folders within the packs/phases. The issue is very consistent across every game I've encountered it with. Always with the same resources that it cant find. I have confirmed its related to the complex model path. Moving everything to the root model-path i.e "resources" in this case. It works flawlessly.

**Comment by rdb** [comment]
Perhaps you can send me (privately or otherwise) a list of all the files you have in the model-path as well as a complete command-line output capture with `notify-level debug` set?

**Comment by thetestgame** [comment]
I will try and setup some example files later this week. I've been extremely busy.

**Comment by rdb** [comment]
Any chance you could post an updated test case?

**Comment by thetestgame** [comment]
I never found time and have since closed the project. If I run across it again ill make another issue. You can probably consider this closed.

**Comment by rdb** [comment]
Closing this issue since there is nothing actionable here.  It can be reopened if a test case or test scenario is supplied.

**Comment by rdb** [comment]
I strongly suspect that this issue may have been fixed by d6b7abedfe9d03c483f1973e34995c9201da46d0, which is a fix I made in response to a similar assertion error caused by a ConfigVariableSearchPath being reloaded by two threads at the same time.  It could indeed also cause some threads to read out an empty search path, which would explain the issue with some files not being found correctly.
